blocks/datasets/__init__.py

```diff
@@ -535,11 +604,72 @@ def get_data(self, request=None):
             self.cache[i] = cache[request:]
         return tuple(data)
 
+    def get_epoch_iterator(self, **kwargs):
+        self.cache = [[] for _ in self.sources]
+        return super(CachedDataStream, self).get_epoch_iterator(**kwargs)
+
     def _cache(self):
         for cache, data in zip(self.cache, next(self.child_epoch_iterator)):
             cache.extend(data)
 
 
+class NGramStream(CachedDataStream):
+    """Return n-grams from a stream.
+
+    This data stream wrapper takes as an input a data stream outputting
+    batches of sentences. From these sentences n-grams of a fixed order
+    (e.g. bigrams, trigrams, etc.) are extracted and returned.
+
+    Parameters
+    ----------
+    ngram_order : int
+        The order of the n-grams to output e.g. 3 for trigrams.
+    data_stream : :class:`DataStream` instance
+        The data stream providing sentences. Each example is assumed to be
+        a list of integers.
+    target_source : str, optional
+        This data stream adds a new source for the target words. By default
```

Comment:
Can you write the the target words is the one next to the n-gram? (as far as I understood from the code)